sympy/sets/fancysets.py

```diff
@@ -501,6 +501,9 @@ def doit(self, **kwargs):
             return FiniteSet(*(f(*a) for a in product(*self.base_sets)))
         return self
 
+    def _kind(self):
+        return self.doit().kind
```

Comment:
This definitely shouldn't call `doit`. Checking the `kind` should always be a cheap operation and `doit` can be arbitrarily expensive.